### PR TITLE
[BEAM-4127] Flink streaming translation for portable pipeline.

### PR DESCRIPTION
**Pull request by tweise**
This is the translator code from the prototype, minus executable stage translation (which I'm going to work an as follow-up).

R: @bsidhom @lukecwik 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `./gradlew build` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+            FlinkExecutionEnvironments.createStreamExecutionEnvironment(
+                    pipelineOptions.as(FlinkPipelineOptions.class));
+    return new StreamingTranslationContext(jobInfo, pipelineOptions, executionEnvironment);
+  }
+
+  /**
+   * Streaming translation context. Stores metadata about known PCollections/DataStreams and holds
+   * the Flink {@link StreamExecutionEnvironment} that the execution plan will be applied to.
+   */
+
```
Comment (lukecwik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java]: nit: whitespace

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+  interface PTransformTranslator<T> {
+    void translate(String id, RunnerApi.Pipeline pipeline, T t);
+  }
+
+  private final Map<String, PTransformTranslator<StreamingTranslationContext>>
+          urnToTransformTranslator = new HashMap<>();
+
+  FlinkStreamingPortablePipelineTranslator() {
+    urnToTransformTranslator.put(PTransformTranslation.FLATTEN_TRANSFORM_URN,
```
Comment (lukecwik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java]: Use ImmutableMap.of for thread safety reasons. HashMap has no concurrency guarantees.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+                                              GlobalWindow.Coder.INSTANCE)));
+      context.addDataStream(Iterables.getOnlyElement(
+              pipeline.getComponents().getTransformsOrThrow(id).getOutputsMap().values()), result);
+    } else {
+      DataStream<T> result = null;
+
+      // Determine DataStreams that we use as input several times. For those, we need to uniquify
+      // input streams because Flink seems to swallow watermarks when we have a union of one and
+      // the same stream.
+      Map<DataStream<T>, Integer> duplicates = new HashMap<>();
```
Comment (lukecwik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java]: Consider using a Multiset (https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Multiset.html) as it provides the frequency count feature already.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+            keyedWorkItemStream
+                    .transform(
+                            pTransform.getUniqueName(),
+                            outputTypeInfo,
+                            (OneInputStreamOperator) doFnOperator);
+
+    context.addDataStream(
+            Iterables.getOnlyElement(pTransform.getOutputsMap().values()),
+            outputDataStream);
+
```
Comment (lukecwik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java]: nit: whitespace

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+                                          StreamingTranslationContext context) {
+    RunnerApi.Components components = pipeline.getComponents();
+    RunnerApi.PTransform transform = components.getTransformsOrThrow(id);
+    RunnerApi.WindowIntoPayload payload;
+    try {
+      payload = RunnerApi.WindowIntoPayload.parseFrom(transform.getSpec().getPayload());
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalArgumentException(e);
+    }
+    WindowFn<T, ? extends BoundedWindow> windowFn = (WindowFn<T, ? extends BoundedWindow>)
```
Comment (lukecwik) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java]: This only works for well known window fns, we should defer this execution to the SDK if the WindowFn can't be parsed or just defer it all the time.

Python already supports WindowInto (https://github.com/apache/beam/blob/5b791489a5884cfd298578324ec08238d28f303c/sdks/python/apache_beam/runners/worker/bundle_processor.py#L556) while Java support is being added in https://github.com/apache/beam/pull/5329

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java (hunk 2 above)
Comment (tweise) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java]: 👍 we don't want to share translator instances between threads but it is cleaner anyways

**Comment by tweise** [comment]
@lukecwik PTAL

**Pull request merged**